File: native/bindings/endstone/permissions/permission_definition.cpp
#include "permission_definition.h"
#include "../../../aegilex_types.h"

#include <aegilex-runtime/src/cxx_host_admin.rs.h>

#include <endstone/permissions/permission.h>
#include <endstone/permissions/permission_default.h>

#include <algorithm>
#include <cstdint>
#include <memory>
#include <string>
#include <string_view>
#include <utility>
#include <vector>

namespace aegilex::native::admin {

class PermissionDefinition::impl {
  public:
    explicit impl(endstone::Permission *definition) noexcept : definition(definition)
    {
    }

    endstone::Permission *definition;
};

namespace {

[[nodiscard]] bool valid_permission_name(const std::string_view name) noexcept
{
    if (name.empty()) {
        return false;
    }
    return std::all_of(name.begin(), name.end(), [](const char value) {
        return (value >= 'a' && value <= 'z') || (value >= '0' && value <= '9') || value == '.' || value == '_' ||
               value == '-';
    });
}

[[nodiscard]] bool permission_default(const std::uint8_t raw, endstone::PermissionDefault &out) noexcept
{
    switch (raw) {
    case aegilex::kPermissionDefaultTrue:
        out = endstone::PermissionDefault::True;
        return true;
    case aegilex::kPermissionDefaultFalse:
        out = endstone::PermissionDefault::False;
        return true;
    case aegilex::kPermissionDefaultOperator:
        out = endstone::PermissionDefault::Operator;
        return true;
    case aegilex::kPermissionDefaultNotOperator:
        out = endstone::PermissionDefault::NotOperator;
        return true;
    case aegilex::kPermissionDefaultConsole:
        out = endstone::PermissionDefault::Console;
        return true;
    default:
        return false;
    }
}

} // namespace

PermissionDefinition::PermissionDefinition(endstone::Permission *definition) noexcept
    : impl(std::make_shared<class PermissionDefinition::impl>(definition))
{
}

endstone::Permission *PermissionDefinition::native() const noexcept
{
    return impl == nullptr ? nullptr : impl->definition;
}

rust::String PermissionDefinition::getName() const
{
    try {
        return native() == nullptr ? rust::String() : rust::String(native()->getName());
    }
    catch (...) {
        return rust::String();
    }
}

rust::String PermissionDefinition::getDescription() const
{
    try {
        return native() == nullptr ? rust::String() : rust::String(native()->getDescription());
    }
    catch (...) {
        return rust::String();
    }
}

void PermissionDefinition::setDescription(const rust::Str description) const
{
    try {
        if (native() != nullptr) {
            native()->setDescription(std::string(description));
        }
    }
    catch (...) {
    }
}

std::uint8_t PermissionDefinition::getDefault() const
{
    try {
        if (native() == nullptr) {
            return aegilex::kPermissionDefaultOperator;
        }
        switch (native()->getDefault()) {
        case endstone::PermissionDefault::True:
            return aegilex::kPermissionDefaultTrue;
        case endstone::PermissionDefault::False:
            return aegilex::kPermissionDefaultFalse;
        case endstone::PermissionDefault::Operator:
            return aegilex::kPermissionDefaultOperator;
        case endstone::PermissionDefault::NotOperator:
            return aegilex::kPermissionDefaultNotOperator;
        case endstone::PermissionDefault::Console:
            return aegilex::kPermissionDefaultConsole;
        }
    }
    catch (...) {
    }
    return aegilex::kPermissionDefaultOperator;
}

void PermissionDefinition::setDefault(const std::uint8_t default_value) const
{
    try {
        endstone::PermissionDefault value = endstone::PermissionDefault::Operator;
        if (native() != nullptr && permission_default(default_value, value)) {
            native()->setDefault(value);
        }
    }
    catch (...) {
    }
}

rust::Vec<PermissionChild> PermissionDefinition::getChildren() const
{
    rust::Vec<PermissionChild> children;
    try {
        if (native() != nullptr) {
            for (const auto &[name, value] : native()->getChildren()) {
                children.push_back(PermissionChild{.name = rust::String(name), .value = value});
            }
        }
    }
    catch (...) {
    }
    return children;
}

void PermissionDefinition::addChild(const rust::Str name, const bool value) const
{
    try {
        if (native() != nullptr && valid_permission_name(std::string_view(name.data(), name.size()))) {
            native()->getChildren()[std::string(name)] = value;
            native()->recalculatePermissibles();
        }
    }
    catch (...) {
    }
}

void PermissionDefinition::removeChild(const rust::Str name) const
{
    try {
        if (native() != nullptr) {
            static_cast<void>(native()->getChildren().erase(std::string(name)));
            native()->recalculatePermissibles();
        }
    }
    catch (...) {
    }
}

void PermissionDefinition::recalculatePermissibles() const
{
    try {
        if (native() != nullptr) {
            native()->recalculatePermissibles();
        }
    }
    catch (...) {
    }
}

std::unique_ptr<PermissionDefinition> PermissionDefinition::addParentByName(const rust::Str name,
                                                                            const bool value) const
{
    try {
        if (native() == nullptr || !valid_permission_name(std::string_view(name.data(), name.size()))) {
            return {};
        }
        auto *parent = native()->addParent(std::string(name), value);
        return parent == nullptr ? std::unique_ptr<PermissionDefinition>()
                                 : std::make_unique<PermissionDefinition>(parent);
    }
    catch (...) {
        return {};
    }
}

PermissionDefinitionCollection::PermissionDefinitionCollection(
    std::vector<std::unique_ptr<PermissionDefinition>> definitions) noexcept
    : definitions_(std::move(definitions))
{
}

std::size_t PermissionDefinitionCollection::len() const noexcept
{
    return definitions_.size();
}

std::unique_ptr<PermissionDefinition>
PermissionDefinitionCollection::takePermissionDefinition(const std::size_t index) noexcept
{
    return index < definitions_.size() ? std::move(definitions_[index]) : std::unique_ptr<PermissionDefinition>();
}

} // namespace aegilex::native::admin

File: rust/crates/aegilex-runtime/src/host/endstone/permissions/permission_definition.rs
//! Core ABI implementation for `native/bindings/endstone/permissions/permission_definition.h`.

use crate::host::endstone::support::*;

impl crate::core_host::imports::HostPermissionDefinition for PluginStoreState {
    fn permission_definition_get_name(
        &mut self,
        self_: u32,
    ) -> Result<Result<String, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permission-definition.permission-definition.get-name")?;
            resolve_definition(
                self,
                definition_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|definition| definition.getName())
            .map_err(map_core_host_error)
        })())
    }

    fn permission_definition_get_description(
        &mut self,
        self_: u32,
    ) -> Result<Result<String, TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "permission-definition.permission-definition.get-description",
            )?;
            resolve_definition(
                self,
                definition_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|definition| definition.getDescription())
            .map_err(map_core_host_error)
        })())
    }

    fn permission_definition_set_description(
        &mut self,
        self_: u32,
        description: String,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "permission-definition.permission-definition.set-description",
            )?;
            resolve_definition(
                self,
                definition_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|definition| definition.setDescription(&description))
            .map_err(map_core_host_error)
        })())
    }

    fn permission_definition_get_default(
        &mut self,
        self_: u32,
    ) -> Result<Result<PermissionDefaultPermissionDefault, TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "permission-definition.permission-definition.get-default",
            )?;
            resolve_definition(
                self,
                definition_handle(self, self_).map_err(map_core_host_error)?,
            )
            .and_then(|definition| permission_default(definition.getDefault()))
            .map_err(map_core_host_error)
        })())
    }

    fn permission_definition_set_default(
        &mut self,
        self_: u32,
        default_value: PermissionDefaultPermissionDefault,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "permission-definition.permission-definition.set-default",
            )?;
            resolve_definition(
                self,
                definition_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|definition| definition.setDefault(native_permission_default(default_value)))
            .map_err(map_core_host_error)
        })())
    }

    fn permission_definition_list_children(
        &mut self,
        self_: u32,
    ) -> Result<Result<Vec<PermissionAttachmentPermissionChild>, TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "permission-definition.permission-definition.list-children",
            )?;
            resolve_definition(
                self,
                definition_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|definition| {
                definition
                    .getChildren()
                    .into_iter()
                    .map(|child| PermissionAttachmentPermissionChild {
                        name: child.name,
                        value: child.value,
                    })
                    .collect()
            })
            .map_err(map_core_host_error)
        })())
    }

    fn permission_definition_add_child(
        &mut self,
        self_: u32,
        child_name: String,
        value: bool,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "permission-definition.permission-definition.add-child",
            )?;
            resolve_definition(
                self,
                definition_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|definition| definition.addChild(&child_name, value))
            .map_err(map_core_host_error)
        })())
    }

    fn permission_definition_remove_child(
        &mut self,
        self_: u32,
        child_name: String,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "permission-definition.permission-definition.remove-child",
            )?;
            resolve_definition(
                self,
                definition_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|definition| definition.removeChild(&child_name))
            .map_err(map_core_host_error)
        })())
    }

    fn permission_definition_recalculate_defaults(
        &mut self,
        self_: u32,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "permission-definition.permission-definition.recalculate-defaults",
            )?;
            let definition = resolve_definition(
                self,
                definition_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map_err(map_core_host_error)?;
            resolve_server(self)
                .map(|server| server.recalculatePermissionDefaults(definition))
                .map_err(map_core_host_error)
        })())
    }

    fn permission_definition_recalculate_permissibles(
        &mut self,
        self_: u32,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "permission-definition.permission-definition.recalculate-permissibles",
            )?;
            resolve_definition(
                self,
                definition_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|definition| definition.recalculatePermissibles())
            .map_err(map_core_host_error)
        })())
    }

    fn permission_definition_add_parent(
        &mut self,
        self_: u32,
        parent_name: String,
        value: bool,
    ) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "permission-definition.permission-definition.add-parent",
            )?;
            let parent = resolve_definition(
                self,
                definition_handle(self, self_).map_err(map_core_host_error)?,
            )
            .map(|definition| definition.addParentByName(&parent_name, value))
            .map_err(map_core_host_error)?;
            self.insert_permission_definition_resource(parent)
                .map_err(map_core_host_error)
        })())
    }

    fn get(&mut self, name: String) -> Result<Result<Option<u32>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permission-definition.get")?;
            let definition = resolve_server(self)
                .map(|server| server.getPermissionDefinition(&name))
                .map_err(map_core_host_error)?;
            (!definition.is_null())
                .then(|| {
                    self.insert_permission_definition_resource(definition)
                        .map_err(map_core_host_error)
                })
                .transpose()
        })())
    }

    fn add(
        &mut self,
        name: String,
        description: Option<String>,
        default_value: Option<PermissionDefaultPermissionDefault>,
        children: Vec<PermissionAttachmentPermissionChild>,
    ) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permission-definition.add")?;
            let (has_description, description) =
                description.map_or((false, String::new()), |value| (true, value));
            let (has_default, default_value) =
                default_value.map_or((false, 0), |value| (true, native_permission_default(value)));
            let children = children
                .into_iter()
                .map(|child| cxx_admin::PermissionChild {
                    name: child.name,
                    value: child.value,
                })
                .collect();
            let definition = resolve_server(self)
                .map(|server| {
                    server.addPermissionDefinition(
                        &name,
                        has_description,
                        &description,
                        has_default,
                        default_value,
                        &children,
                    )
                })
                .map_err(map_core_host_error)?;
            self.insert_permission_definition_resource(definition)
                .map_err(map_core_host_error)
        })())
    }

    fn remove(&mut self, name: String) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permission-definition.remove")?;
            let removed = resolve_server(self)
                .map(|server| server.removePermissionDefinitionByName(&name))
                .map_err(map_core_host_error)?;
            if !removed {
                return Err(TypesHostError::NotFound);
            }
            let handles = self
                .handles
                .permission_definitions_named(self.invocation_id, &name);
            self.handles.remove_handles(&handles);
            Ok(())
        })())
    }

    fn list_defaults(
        &mut self,
        level: PermissionLevelPermissionLevel,
    ) -> Result<Result<Vec<u32>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "permission-definition.list-defaults")?;
            let level = match level {
                PermissionLevelPermissionLevel::Default => 0,
                PermissionLevelPermissionLevel::Operator => 1,
                PermissionLevelPermissionLevel::Console => 2,
            };
            let mut definitions = resolve_server(self)
                .map(|server| server.listDefaultPermissionDefinitions(level))
                .map_err(map_core_host_error)?;
            let len = definitions.len();
            (0..len)
                .map(|index| {
                    self.insert_permission_definition_resource(
                        definitions.pin_mut().takePermissionDefinition(index),
                    )
                    .map_err(map_core_host_error)
                })
                .collect()
        })())
    }
}
